### PR TITLE
10442 response csv export modal

### PR DESCRIPTION
**Pull request by cooperka**
Screen Shot 2020-02-26 at 15 50 33

Updated hint wording:

Screen Shot 2020-03-03 at 12 02 30

### CODE REVIEW

**Comment by cooperka** [comment]
Spec failure due to `\n` vs `\r\n` in fixture file, I fixed it locally and committed but I think git automatically converted it. Will look into better conversion tomorrow.

```
  1) Results::Csv::Generator with long_text_behavior exports truncated values
     Failure/Error: is_expected.to match_user_facing_csv(prepare_response_csv_expectation("truncated_values.csv"))
       expected "﻿ResponseID,Shortcode,Form,Submitter,DateSubmitted,Reviewed,GroupName,GroupLevel,TextQ1,TextQ2\r\n7b...866-31d4612fca00,vv-oqv-ql8g2,Sample Form 1,A User 1,2015-11-20 06:30:00-06,false,,0,foo,foo ba\r\n"
       to match user facing csv "ResponseID,Shortcode,Form,Submitter,DateSubmitted,Reviewed,GroupName,GroupLevel,TextQ1,TextQ2\n7b8ad...-9866-31d4612fca00,vv-oqv-ql8g2,Sample Form 1,A User 1,2015-11-20 06:30:00-06,false,,0,foo,foo ba\n"
       Diff:
       @@ -1,3 +1,3 @@
       -ResponseID,Shortcode,Form,Submitter,DateSubmitted,Reviewed,GroupName,GroupLevel,TextQ1,TextQ2
       +﻿ResponseID,Shortcode,Form,Submitter,DateSubmitted,Reviewed,GroupName,GroupLevel,TextQ1,TextQ2
        7b8ad3ee-fe20-4262-9866-31d4612fca00,vv-oqv-ql8g2,Sample Form 1,A User 1,2015-11-20 06:30:00-06,false,,0,foo,foo ba
     # ./spec/models/results/csv/generator_spec.rb:315:in `block (3 levels) in <top (required)>'
```

File: spec/support/matchers/matchers.rb [hunk 1]
```diff
@@ -70,7 +70,7 @@ def actual_row(report, row, row_idx)
 RSpec::Matchers.define(:match_user_facing_csv) do |expected|
   match do |actual|
     doc = "\xEF\xBB\xBF" + expected
-    doc.to_s == actual
+    doc.to_s.gsub("\r\n", "\n") == actual.gsub("\r\n", "\n")
```
Comment (smoyth) [inline spec/support/matchers/matchers.rb]: This sets off some alarm bells for me. Why is this needed? I'd prefer to match line endings exactly. We need to have control over those for reasons mentioned above. It was not an accident that we had non-standard line endings in some of those CSV files before... Help me understand what happened here?

File: spec/models/results/csv/answer_processor_spec.rb [hunk 2]
```diff
@@ -125,4 +125,75 @@
...
+
+    it "truncates after max newlines (above max)" do
+      expect(buffer).to receive(:write).with("Q1", "12\r\n34")
+      processor.process(
+        {"question_code" => "Q1", "value" => +"12\n34\n5"},
+        long_text_behavior: "truncate"
+      )
+    end
+
+    it "counts each /r/n as a single line" do
```
[the remark this answers is not in the document]
Comment (cooperka) [reply spec/models/results/csv/answer_processor_spec.rb]: Escaping `\\r\\n` technically should work but it broke my test debugger 💩 I will change it to `CRLF` instead.

File: spec/support/matchers/matchers.rb (hunk 1 above)
Comment (cooperka) [reply spec/support/matchers/matchers.rb]: Basically these specs pass locally but fail on Travis without the `gsub` here. I am unable to actually commit proper newlines because git converts them automatically.

It may be configurable with a local git config, I will look into this. I didn't realize the endings were so critical to Excel, so this clearly isn't a good solution.

File: spec/fixtures/response_csv/option_values.csv [hunk 3]
```diff
@@ -1,2 +1,2 @@
-ResponseID,Shortcode,Form,Submitter,DateSubmitted,Reviewed,GroupName,GroupLevel,SelectMultipleQ2,SelectOneQ1
-*id1*,*shortcode1*,Sample Form 1,A User 1,2015-11-20 06:30:00-06,false,,0,5;8,2
+ResponseID,Shortcode,Form,Submitter,DateSubmitted,Reviewed,GroupName,GroupLevel,SelectMultipleQ2,SelectOneQ1
+*id1*,*shortcode1*,Sample Form 1,A User 1,2015-11-20 06:30:00-06,false,,0,5;8,2
```
[the remark this answers is not in the document]
Comment (cooperka) [reply spec/fixtures/response_csv/option_values.csv]: See https://github.com/thecartercenter/nemo/pull/717#discussion_r385905398, will find a better solution.

File: spec/fixtures/option_set_imports/simple.csv [hunk 4]
```diff
@@ -1,27 +1,27 @@
...
+Nord-Kivu,,,,
+Sud-Kivu,,,,
+Kwango,,,,
+Kwilu,,,,
+Mai-Ndombe,,,,
+Kongo Central,,,,
+Kinshasa,,,,
+Bas-Uele,,,,
+Haut-Uele,,,,
+Ituri,,,,
```
[the remark this answers is not in the document]
Comment (cooperka) [reply spec/fixtures/option_set_imports/simple.csv]: See https://github.com/thecartercenter/nemo/pull/717#discussion_r385905398, will find a better solution.

File: app/jobs/response_csv_export_operation_job.rb [hunk 5]
```diff
@@ -2,9 +2,10 @@
 
 # Operation for exporting response CSV.
 class ResponseCsvExportOperationJob < OperationJob
-  def perform(operation, search: nil)
+  def perform(operation, search: nil, export_options: nil)
     ability = Ability.new(user: operation.creator, mission: mission)
-    result = generate_csv(responses(ability, search))
+    options_to_splat = export_options&.symbolize_keys || {}
+    result = generate_csv(responses(ability, search), **options_to_splat)
```
[the remark this answers is not in the document]
Comment (cooperka) [reply app/jobs/response_csv_export_operation_job.rb]: Hmm, the idea behind this was to allow `answer_processor` to accept a named param `long_text_behavior` instead of parsing `options`, which your suggestion still accomplishes 🎉 I cleaned this up, thanks.

File: app/models/response_csv_export_options.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+# Custom options for exporting responses to CSV.
+class ResponseCSVExportOptions
+  include ActiveModel::Model
+
+  attr_accessor :long_text_behavior
+
+  LONG_TEXT_BEHAVIOR_OPTIONS = %w[exclude truncate include].freeze
+
+  def initialize(**attribs)
```
[the remark this answers is not in the document]
Comment (cooperka) [reply app/models/response_csv_export_options.rb]: Fixed this in some other locations as well.

File: app/models/results/csv/answer_processor.rb [hunk 7]
```diff
@@ -109,6 +116,36 @@ def convert_mac_line_endings_to_windows(str)
...
+
+      # Return a string that's been modified according to long_text_behavior.
+      def handle_long_text(str)
+        return str if long_text_behavior == "include"
+        exclude = long_text_behavior == "exclude" # Otherwise truncate.
+
+        chars = str.length
+        if chars > MAX_CHARACTERS
+          return "" if exclude
+          str = str[0, MAX_CHARACTERS].sub(/\r$/, "") # Don't allow a dangling "\r" with no "\n".
```
[the remark this answers is not in the document]
Comment (cooperka) [reply app/models/results/csv/answer_processor.rb]: A function that ONLY trims ENDING newlines, how convenient! 🙌

File: app/models/results/csv/answer_processor.rb [hunk 8]
```diff
@@ -69,21 +74,23 @@ def write_select_cells
...
           break
         end
       end
 
-      def convert_line_endings(str)
+      def normalize_value(str)
         # We do this with loops instead of regexps b/c regexps are slow.
         convert_unix_line_endings_to_windows(str)
         convert_mac_line_endings_to_windows(str)
+        handle_long_text(str)
```
[the remark this answers is not in the document]
Comment (cooperka) [reply app/models/results/csv/answer_processor.rb]: All methods now act in place ⚡️ thank you for pointing out some useful methods.

File: config/locales/en/main.yml [hunk 9]
```diff
@@ -670,6 +671,8 @@ en:
         bar_style: "Whether bars should be stacked or side-by-side."
       response:
         incomplete: "Check this box if there are required questions for which you can't provide answers. You will then be able to save this response without answering those questions and this response will be marked as 'incomplete'."
+      response_csv_export_options:
+        long_text_behavior: "Some programs limit the number of characters and lines each data cell can contain. You can optionally modify the exported data as follows:\n\n\n\n&nbsp;&nbsp;&nbsp;<b>Exclude</b>: If an answer hits a limit, that answer will exported as blank.\n\n&nbsp;&nbsp;&nbsp;<b>Truncate</b>: If an answer hits a limit, it will be cut off there. This may break HTML contained in the exported answer.\n\n&nbsp;&nbsp;&nbsp;<b>Include</b>: No modification will be made."
```
[the remark this answers is not in the document]
Comment (cooperka) [reply config/locales/en/main.yml]: A few other strings use the same format `\n\n&nbsp;&nbsp;&nbsp;` for indenting example data, such as code blocks and answers. No other strings use newlines at all besides those.

File: spec/support/matchers/matchers.rb (hunk 1 above)
Comment (cooperka) [reply spec/support/matchers/matchers.rb]: Fixed with `git config --local core.autocrlf false`; I also rebased and excluded the old commits from history.

File: spec/models/results/csv/generator_spec.rb [hunk 10]
```diff
@@ -298,6 +300,22 @@
...
+
+    before do
+      stub_const(Results::Csv::AnswerProcessor, "MAX_CHARACTERS", 6)
+      Timecop.freeze(submission_time) do
+        create_response(form: form, answer_values: ["foo", "foo bar baz"])
+      end
+    end
+
+    it "exports truncated values" do
+      is_expected.to match_user_facing_csv(prepare_response_csv_expectation("truncated_values.csv"))
```
[the remark this answers is not in the document]
Comment (smoyth) [reply spec/models/results/csv/generator_spec.rb]: Lol I meant testing pyramid 😜
